### PR TITLE
Fix implementation of allowEmptyItems in QuietFactory

### LINKED ISSUE #230: bug in QuietFactory
the quiet factory seems to mess up on the allowEmptyItems option: https://github.com/symfony-cmf/MenuBundle/blob/master/QuietFactory.php#L57
we implicitly return null as well. the idea of allowing empty items would be to return a menu item with just the text but no URL.

### PR DESCRIPTION
**Pull request by WouterJ**
Fixes #230 

While fixing that bug, I also discovered some flaws with the logging, which I also fix in this PR.

### CODE REVIEW

**Comment by WouterJ** [comment]
Tests fail because of a bug in Symfony: https://github.com/symfony/symfony/pull/15619

File: Tests/Unit/QuietFactoryTest.php [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+        $this->innerFactory = $this->prophesize('Knp\Menu\FactoryInterface');
+        $this->logger = $this->prophesize('Psr\Log\LoggerInterface');
+    }
+
+    public function testAllowEmptyItemsReturnsItemWithoutURL()
+    {
+        $this->innerFactory->createItem('Home', array('route' => 'not_existent'))
+            ->willThrow('Symfony\Component\Routing\Exception\RouteNotFoundException');
+
+        $this->innerFactory->createItem('Home', array())->shouldBeCalled();
```
Comment (stof) [inline Tests/Unit/QuietFactoryTest.php]: a willReturn and asserting the return value of ``$factory->createItem`` would be better than a ``shouldBeCalled`` assertion: the important part is not that the inner factory is called, but that the item is returned

Comment (stof) [inline Tests/Unit/QuietFactoryTest.php]: btw, it may make sense to add another test for ``allowEmptyItems`` being false

File: QuietFactory.php [hunk 2]
```diff
@@ -49,11 +48,21 @@ public function createItem($name, array $options = array())
...
+                    array('exception' => $e)
+                );
+            }
 
             if (!$this->allowEmptyItems) {
                 return null;
             }
+
+            // remove route to create an item without any URL
+            unset($options['route']);
```
Comment (dbu) [inline QuietFactory.php]: could it be something else like the content of the menu item that is RouteReferrersReadInterface but does not have any route associated? or some random content that does not even have the interface? we should also remove other things that our factory might be looking at.

File: QuietFactory.php [hunk 3]
```diff
@@ -49,11 +48,21 @@ public function createItem($name, array $options = array())
         try {
             return $this->innerFactory->createItem($name, $options);
         } catch (RouteNotFoundException $e) {
-            $this->logger->error(sprintf('%s : %s', $name, $e->getMessage()));
+            if (null !== $this->logger) {
```
Comment (dbu) [inline QuietFactory.php]: what about instantiating a NullLogger in the constructor if the logger was null?

**Comment by dbu** [comment]
:+1:
i guess we should wait for the symfony bugfix to be merged before we merge this?

**Comment by aitboudad** [comment]
it should be fixed now

File: QuietFactory.php (hunk 3 above)
Comment (WouterJ) [inline QuietFactory.php]: This was rejected on Core many times, for performance reasons.

File: QuietFactory.php (hunk 2 above)
Comment (WouterJ) [inline QuietFactory.php]: Hmm, you're correct about the content.

What about reverting this unset and setting linkType to `uri` instead? This means that the route option will be unset by the ContentExtension of this bundle and the content object (if available) will still be stored in the extra's.

File: QuietFactory.php (hunk 3 above)
Comment (dbu) [inline QuietFactory.php]: okay then lets leave it as is

**Comment by dbu** [comment]
@WouterJ now the twig deprecations hit us here as well. do you want to add the deprecation env variable to 2.3 and 2.6 as well?

and i guess for the next version after this one, we should drop 2.3 support and actually fix all those deprecations. also requires the sonata upgrade.

**Comment by WouterJ** [comment]
Updated the PR.

> and i guess for the next version after this one, we should drop 2.3 support and actually fix all those deprecations. also requires the sonata upgrade.

Yes, that's what we decided to do when Symfony 3.0 is released.

**Comment by lsmith77** [comment]
needs a rebase

**Comment by WouterJ** [comment]
Rebased

**Pull request merged**